com.dynamo.cr/com.dynamo.cr.bob/src/com/dynamo/bob/archive/publisher/ZipPublisher.java

```diff
@@ -118,19 +115,15 @@ public void publish(ArchiveEntry entry, InputStream data) throws CompileExceptio
         final String archiveEntryHexdigest = entry.getHexDigest();
         final String archiveEntryName = entry.getName();
         final String zipEntryName = (archiveEntryHexdigest != null) ? archiveEntryHexdigest : archiveEntryName;
-        synchronized (zipEntries) {
-            if (zipEntries.contains(zipEntryName)) {
-                return;
-            }
-            zipEntries.add(zipEntryName);
+        if (entries.put(archiveEntryName, entry) != null) {
```

Comment:
Simpler logic when using `ConcurrentHashMap`
also, no need to check contains and add it separatly